### PR TITLE
feat: generate scheme.d.ts automatically - close #393

### LINKED ISSUE #393: add `schema.d.ts`?
As title?

### PR DESCRIPTION
**Pull request by JounQin**
fixes #393

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,6 +23,8 @@
   "author": "Gregor Martynus (https://twitter.com/gr2m)",
   "license": "MIT",
   "devDependencies": {
+    "@graphql-codegen/cli": "^1.19.2",
+    "@graphql-codegen/typescript": "^1.18.0",
```
Comment (gr2m) [inline package.json:27]: I thin we should also add a `"types"` key to the package.json as per https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

File: bin/download.js [hunk 2]
```diff
@@ -38,5 +38,8 @@ request("/graphql")
       return;
     }
 
+    console.log("⌛  Generating schema.d.ts …");
+    await execa("npx", ["graphql-codegen"]);
+
```
Comment (gr2m) [inline bin/download.js:43]: I think we need to separate the download from updating the `schema.d.ts` file. The `.d.ts` can change even if no update exists in GiHub's schema due to updates in one of the `@graphql-codegen/*` packages.

I would suggest to do the following:

1. Create a new `bin/generate-typescript.js` file
2. In scripts, rename the current `"update"` script to `"download"`
3. Add a new `"generate-typescript"` script
4. Add a new `"update"` script which does `"npm run -s download && npm run -s generate-typescript"`

Then add a new workflow file `.github/workflows/update-graphql-codegen.yml` which runs on

```yml
"on":
  push:
    branches:
      - dependabot/npm_and_yarn/graphql-codegen/*
```

and updates the existing pull request from dependabot if it exists, compare `.github/workflows/update-prettier.yml`

File: codegen.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+overwrite: true
+schema: "schema.graphql"
+generates:
+  schema.d.ts:
+    plugins:
+      - "typescript"
```
Comment (gr2m) [inline codegen.yml:6]: Is there way to add the configuration inline, or add it to `package.json`? I would prefer not to add new config files to the repository root.

File: README.md [hunk 4]
```diff
@@ -27,6 +29,34 @@ schema.json; // JSON version
...
+      }
+    }
+  `,
+  {
+    headers: {
+      authorization: `token secret123`,
+    },
+  }
+);
+```
```
Comment (gr2m) [inline README.md:58]: # 🤯 

this is awesome!

**Review by gr2m — COMMENTED** [review]
this is super dope, I didn't know this was possible! I've a few comments, let me know if you have any questions, or if you'd like me to finish up the PR on my own. 

Thanks again!

File: package.json (hunk 1 above)
Comment (JounQin) [reply package.json:27]: We've already had `types` field for `index.d.ts`, so I `export * from './scheme'` in `index.d.ts`

File: codegen.yml (hunk 3 above)
Comment (JounQin) [reply codegen.yml:6]: It seems not supported now.

Comment (JounQin) [reply codegen.yml:6]: Wait for https://github.com/dotansimha/graphql-code-generator/pull/5144 to be released.

File: bin/generate-typescript.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+const execa = require("execa");
+
+console.log("⌛  Generating schema.d.ts …");
+
+execa("npx", ["graphql-codegen"]);
```
Comment (gr2m) [inline bin/generate-typescript.js:5]: I wonder if it makes sense to create a JS file for it, if all it does is to run `graphql-codegen`? I think I'd instead update the script in package.json to

```
     "generate-typescript": "graphql-codegen",
```

**Review by gr2m — APPROVED** [review]
Thanks so much, this is really great!

**Pull request merged**